lib/nx_stencil.ex

```diff
@@ -70,40 +84,65 @@ defmodule NxStencil do
     worker(parent, rank, up, down, top_halo, data, bottom_halo, iters - 1)
   end
 
-  def process_start do
+  defp process_start do
     receive do
-      {parent, :start_data,
-       {
-         rank,
-         up,
-         down,
-         top_halo,
-         init_data,
-         bottom_halo,
-         iters
-       }} ->
+      %InitData{
+        parent: parent,
+        data: %StencilData{
+          up_rank: up,
+          self_rank: rank,
+          down_rank: down,
+          top_halo: top_halo,
+          self_data: init_data,
+          bottom_halo: bottom_halo
+        },
+        iters: iters
+      } ->
         worker(parent, rank, up, down, top_halo, init_data, bottom_halo, iters)
     end
   end
 
-  def get_data(x_dim, y_dim) do
+  defp all_not_nil(lst) do
+    lst
+    |> Stream.map(fn x -> x != nil end)
+    |> Enum.reduce(true, fn x, acc -> x and acc end)
+  end
+
+  defp collect_impl(list) do
+    # Can't call function in guard?
+    if all_not_nil(list) do
+      list
+    else
+      receive do
+        {:end_data, rank, data} -> collect_impl(list |> List.replace_at(rank, data))
+      end
+    end
+  end
+
+  defp collect(count) do
+    collect_impl(List.duplicate(nil, count))
+  end
+
+  defp get_data(x_dim, y_dim) do
     # 2d array of incrementing numbers from 0 to x*y
     # Nx.tensor(Enum.chunk_every(0..(x_dim * y_dim - 1), y_dim), type: :f64)
     Enum.chunk_every(0..(x_dim * y_dim - 1), y_dim)
   end
 
-  def run() do
+  defp run_impl() do
     # For now, process count == row count
-    proc_count = 5
-    iters = 10
+    proc_count = 64
+    iters = 100
+    x_dim = proc_count
+    y_dim = 1000
     IO.puts("start")
-    data = NxStencil.get_data(proc_count, 10)
+    data = get_data(x_dim, y_dim)
 
     procs =
       for rank <- 0..(proc_count - 1) do
         {
           rank,
-          spawn(&NxStencil.process_start/0)
+          spawn(&process_start/0)
```

Comment:
Using spawn is the primitive. This could be simplified using `Task` and `async_stream`. Then, you don't have to do the send. It will exit and return the value.